vm/devices/storage/disk_nvme/nvme_driver/fuzz/fuzz_nvme_driver.rs

```diff
@@ -0,0 +1,169 @@
+use pal_async::DefaultDriver;
+use pci_core::msi::MsiInterruptSet;
+use scsi_buffers::OwnedRequestBuffers;
+use std::sync::Arc;
+use user_driver::emulated::DeviceSharedMemory;
+use vmcore::vm_task::SingleDriverBackend;
+use vmcore::vm_task::VmTaskDriverSource;
+
+/// Nvme driver fuzzer
+pub struct FuzzNvmeDriver {
+    driver: Option<NvmeDriver<FuzzEmulatedDevice<NvmeController>>>,
+    namespace: Namespace,
+    payload_mem: GuestMemory,
+}
+
+impl FuzzNvmeDriver {
+    /// Setup a new nvme driver with a fuzz-enabled backend device.
+    pub async fn new(driver: DefaultDriver) -> Self {
+        // Physical storage to back the disk
+        let ram_disk = RamDisk::new(1 << 20, false).unwrap();
+
+        let base_len = 64 << 20;  // 64MB
+        let payload_len = 1 << 20;  // 1MB
+        let mem = DeviceSharedMemory::new(base_len, payload_len);
+
+        // Trasfer buffer
+        let payload_mem = mem
+            .guest_memory()
+            .subrange(base_len as u64, payload_len  as u64, false)
+            .unwrap();
+
+        // Nvme device and driver setup
+        let driver_source =
+            VmTaskDriverSource::new(SingleDriverBackend::new(driver));
+        let mut msi_set = MsiInterruptSet::new();
+        let nvme = NvmeController::new(
+            &driver_source,
+            mem.guest_memory().clone(),
+            &mut msi_set,
+            &mut ExternallyManagedMmioIntercepts,
+            NvmeControllerCaps {
+                msix_count: 2,
+                max_io_queues: 64,
+                subsystem_id: Guid::new_random(),
+            },
+        );
+
+        nvme.client()
+            .add_namespace(1, Arc::new(ram_disk))
+            .await
+            .unwrap();
+
+        let device = FuzzEmulatedDevice::new(nvme, msi_set, mem);
+        let nvme_driver = NvmeDriver::new(&driver_source, 64, device).await.unwrap();
+        let namespace = nvme_driver.namespace(1).await.unwrap();
+
+        Self {
+            driver: Some(nvme_driver),
+            namespace,
+            payload_mem,
+        }
+    }
+
+    /// Clean up fuzzing infrastructure.
+    pub async fn shutdown(&mut self) {
+        self.namespace
+            .deallocate(
+                0,
+                &[
+                    DsmRange {
+                        context_attributes: 0,
+                        starting_lba: 1000,
+                        lba_count: 2000,
+                    },
+                    DsmRange {
+                        context_attributes: 0,
+                        starting_lba: 2,
+                        lba_count: 2,
+                    },
+                ],
+            )
+            .await
+            .unwrap();
+
+        self.driver.take().unwrap().shutdown().await;
+    }
+
+    /// Generates and executes an arbitrary NvmeDriverAction. Returns either an arbitrary error or the executed action.
+    pub async fn execute_arbitrary_action(&mut self) -> Result<NvmeDriverAction, arbitrary::Error> {
```

Comment:
Does this need to return the action still? Doesn't hurt i suppose.

Reply:
We need a way to signal the that the fuzzer is done. 2 options:
* Return an error if we were unable to create an action and break.
* Upon each iteration of the loop, check how many bytes are left in the RAW data. But in this case if there are 10 bytes left, we would have to account for a minimum number of bytes that is needed by the `execute_arbitrary_action` function (which could also change).

First approach seemed cleaner imo.

Reply:
What I meant is it can still return the Error, but the action isn't being used anywhere. Typically in cases like that the return type will be Result<(), Error> or similar.

Reply:
Updated now.